types/query/pagination_test.go

```diff
@@ -0,0 +1,174 @@
+	require.NoError(t, err)
+	require.Equal(t, res.Balances.Len(), defaultLimit)
+	require.NotNil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with custom limit, key and countTotal false")
+	pageReq = &query.PageRequest{Key: res.Res.NextKey, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.Equal(t, res.Balances.Len(), defaultLimit)
+	require.NotNil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate for last page, results in records less than max limit")
+	pageReq = &query.PageRequest{Key: res.Res.NextKey, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), defaultLimit)
+	require.Equal(t, res.Balances.Len(), lastPageRecords)
+	require.Nil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with offset and limit")
+	pageReq = &query.PageRequest{Offset: 200, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), defaultLimit)
+	require.Equal(t, res.Balances.Len(), lastPageRecords)
+	require.Nil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with offset and limit")
+	pageReq = &query.PageRequest{Offset: 100, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), defaultLimit)
+	require.NotNil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with offset and key - error")
+	pageReq = &query.PageRequest{Key: res.Res.NextKey, Offset: 100, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.Error(t, err)
+	require.Equal(t, err.Error(), "invalid request, either offset or key is expected, got both")
+
+	t.Log("verify paginate with offset greater than total results")
+	pageReq = &query.PageRequest{Offset: 300, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), 0)
+	require.Nil(t, res.Res.NextKey)
+}
+
+func SetupTest(t *testing.T) (*simapp.SimApp, sdk.Context) {
```

Comment:
No need to make this public:

```suggestion
func setupTest(t *testing.T) (*simapp.SimApp, sdk.Context) {
```